File: Python-codes/naive_bayes.py
from sklearn.datasets import load_iris # data set
from sklearn.model_selection import train_test_split # data set splitting
from sklearn.naive_bayes import GaussianNB # model fitting
from sklearn.metrics import classification_report, confusion_matrix # accuracy and confusion matrix
import seaborn as sns # confusion matrix
import matplotlib.pyplot as plt # confusion matrix

# data set
iris = load_iris()

# data manipulation
X = iris.data
y = iris. target

# data set splitting
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 42)

# model fitting
naive_b = GaussianNB().fit(X_train, y_train)

# y_pred
y_pred = naive_b.predict(X_test)

# accuracy
print(classification_report(y_test, y_pred))

# confusion matrix
conf_mat = confusion_matrix(y_test, y_pred)
sns.heatmap(conf_mat, annot = True, cmap = 'Blues', fmt = 'g')
plt.xlabel('Predicted')
plt.ylabel('Actual')
plt.show()

File: R-codes/naive_bayes.R
library(caret)      # train/test split, confusionMatrix
library(e1071)      # Naive Bayes
library(ggplot2)    # visualization
library(reshape2)   # confusion matrix reshaping

# data manipulation
X <- iris[, 1:4]      # features
y <- iris$Species     # target

set.seed(12)

# data set splitting (80% train, 20% test)
train_index <- createDataPartition(y, p = 0.8, list = FALSE)

X_train <- X[train_index, ]
X_test  <- X[-train_index, ]

y_train <- y[train_index]
y_test  <- y[-train_index]

# model fitting (gaussian naive bayes)
naive_b <- naiveBayes(x = X_train, y = y_train)

# prediction
y_pred <- predict(naive_b, X_test)

# confusion matrix
confusionMatrix(y_pred, y_test)

# confusion matrix (heatmap)
conf_mat <- table(Actual = y_test, Predicted = y_pred)
conf_df <- melt(conf_mat)

ggplot(conf_df, aes(x = Predicted, y = Actual, fill = value)) +
  geom_tile() +
  geom_text(aes(label = value), color = "white", size = 5) +
  scale_fill_gradient(low = "#ffd9a1", high = "#d47f00") +
  labs(title = "Confusion Matrix",
       x = "Predicted",
       y = "Actual") +
  theme_minimal()
